std/range/package.d

```diff
@@ -1574,18 +1574,20 @@ Returns:
     alias of that range's type.
  */
 auto chooseAmong(Ranges...)(size_t index, Ranges rs)
-if (Ranges.length > 2
-        && is(typeof(choose(true, rs[0], rs[1])))
-        && is(typeof(chooseAmong(0, rs[1 .. $]))))
+if (Ranges.length >= 2 && allSatisfy!(isInputRange, staticMap!(Unqual, Ranges)))
 {
-    return choose(index == 0, rs[0], chooseAmong(index - 1, rs[1 .. $]));
-}
-
-/// ditto
-auto chooseAmong(Ranges...)(size_t index, Ranges rs)
-if (Ranges.length == 2 && is(typeof(choose(true, rs[0], rs[1]))))
-{
-    return choose(index == 0, rs[0], rs[1]);
+    static if (Ranges.length == 2 && is(typeof(choose(true, rs[0], rs[1]))))
+    {
+        return choose(index == 0, rs[0], rs[1]);
+    }
+    else static if (is(typeof(choose(true, rs[0], rs[1]))) && is(typeof(chooseAmong(0, rs[1 .. $]))))
+    {
+        return choose(index == 0, rs[0], chooseAmong(index - 1, rs[1 .. $]));
+    }
+    else
+    {
+        static assert(0, "The ranges have no common type.");
```

Comment:
I agree 100% with @quickfur's comment, but in this case requiring that there's a common element type between all ranges is part of the intent, not just an implementation detail.
Unless we decide to return a range of `Algebraic!(E1, E2, ...)` elements, I don't think this will change in the future.